Google Ads: Add tracking

itsmeichigo opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #13273 
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR adds tracking events for the Google Ads integration in the web view.

Event | Trigger | Properties
-- | -- | --
googleads_entry_point_displayed | When the Google for Woo entry point is shown to the user. | source
googleads_entry_point_tapped | When the Google for Woo entry point is tapped by the user and opens to campaign creation. | source, type, has_campaigns
googleads_flow_started | When the Google for Woo webview is first loaded. | source
googleads_flow_canceled | When the Google for Woo webview is dismissed without completing the flow. | source
googleads_campaign_creation_success | When the Google for Woo webview is automatically dismissed after completing the creation | source
googleads_flow_error | When the Google for Woo webview returns an error. | source

To support tracking web flow errors, `DefaultAuthenticatedWebViewModel` and `WebViewHostingController` has been updated with a new closure to be triggered when loading a web page fails.

## Testing steps
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->
- Pre-requisite: Install, activate, and set up the Google Ads plugin to a test site following pe5sF9-2Qo-p2. Ensure that you're using the plugin version 2.7.5 or above.
- Build the app and log in to your store.
- Enable Google Ads card on the dashboard if you haven't already.
- Confirm that `googleads_entry_point_displayed` is tracked with `source: my_store`.
- Tap on either the Create campaign button or existing campaign detail, and confirm that `googleads_entry_point_tapped` is tracked with `source: my_store` and the correct value for `has_campaigns` depending on whether your store has existing campaigns.
- Switch to the menu tab, confirm that `googleads_entry_point_displayed` is tracked with `source: more_menu`.
- Tap on the Google for WooCommerce item, confirm that `googleads_entry_point_tapped` is tracked with `source: my_store` and the correct value for `has_campaigns` depending on whether your store has existing campaigns.
- When the web view first loaded, confirm that `googleads_flow_started` is tracked with the correct `source` property.
- Dismiss the web view, confirm that `googleads_flow_canceled` is tracked with the correct `source` property.
- Create a campaign on the web view. After the creation completes, confirm that `googleads_campaign_creation_success` is tracked with the correct `source` property.
- Turn off the internet connection and tap on the entry point again. After the loading fails, confirm that `googleads_flow_error` is tracked with the correct `source` property and error details.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->
N/A

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

itsmeichigo committed:
Add new event names for google ads

itsmeichigo committed:
Add helper methods to track new events for google ads

itsmeichigo committed:
Add source for `GoogleAdsCampaignCoordinator`

itsmeichigo committed:
Track entry point tap

itsmeichigo committed:
track flow canceled

itsmeichigo committed:
Track campaign creation success

itsmeichigo committed:
Track entry point displayed for hub menu

itsmeichigo committed:
Track google ads entry point displayed

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  1 Message: <p>This PR contains changes to...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false"><p>This PR contains changes to Tracks-related logic. Please ensure (<strong>author and reviewer</strong>) the following are completed:</p>

<ul>
  <li>The tracks events must be validated in the Tracks system.</li>
  <li>Verify the internal Tracks spreadsheet has also been updated.</li>
  <li>Please consider registering any new events.</li>
  <li>The PR must be assigned the <strong>category: tracks</strong> label.</li>
</ul></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

itsmeichigo committed:
Handle tracking flow start and error

itsmeichigo committed:
Update tracking for entry point displayed

selanthiraiyan commented on WooCommerce/Classes/GoogleAds/GoogleAdsCampaignCoordinator.swift:
```diff
@@ -55,23 +66,38 @@ private extension GoogleAdsCampaignCoordinator {
     func createCampaignViewController(with url: URL) -> UIViewController {
         let redirectHandler: (URL) -> Void = { [weak self] newURL in
+            guard let self else { return }
+            if !hasTrackedStartEvent {
+                ServiceLocator.analytics.track(event: .GoogleAds.flowStarted(source: source))
+                hasTrackedStartEvent = true
+            }
```
❓ Would it be possible to track this event only after the user successfully logs into the web view? This way, we don't end up counting users who skipped login in the funnel while measuring users who started the flow but didn't finish by creating a campaign.

itsmeichigo replied:
This makes sense! Also I think we should only track this event after the page is done loaded. I updated the logic for this in 47627c8c16ebd38afa9829025321293b4d70a6f0.

selanthiraiyan replied:
Thanks for the updates, Huong! 👍

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Dashboard/GoogleAds/GoogleAdsDashboardCardViewModel.swift:
```diff
@@ -79,10 +85,27 @@ final class GoogleAdsDashboardCardViewModel: ObservableObject {
+        $canShowOnDashboard.removeDuplicates()
+            .combineLatest($syncingData.removeDuplicates(), $viewAppeared)
+            .filter { canShow, syncingData, viewAppeared in
+                return canShow && !syncingData && viewAppeared
```
Super nit: Kindly consider adding a comment to explain this filter.

itsmeichigo replied:
Added comments in 7c4a27c9ca88a4183678c4c3053c0afea6e050d3.

selanthiraiyan reviewed (approved):
The tracking works as expected. 🚀

itsmeichigo committed:
Add comments for the entry point displayed event

itsmeichigo committed:
Update tracking for the start of the web flow

itsmeichigo committed:
Track flow type for entryPointTapped event

itsmeichigo commented:
Thanks @selanthiraiyan for the reviews! I added changes to the tracking of the start of the web flow as you suggested. I also added a new property to track the flow type of the entry point tap event in fa6bb27a1921684d7c16f6d997451564f20eb510. Please feel free to take another look when you can, I'll merge this PR later today.

itsmeichigo committed:
Update flow type tracking value

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI Components/WebView.swift:
```diff
@@ -54,12 +58,18 @@ struct WebView: UIViewRepresentable {
     /// A url to trigger dismissing of the web view.
     let urlToTriggerExit: String?
 
+    /// Optional closure for when a web page loads successfully.
+    let pageLoadHandler: ((URL) -> Void)?
```
Super nit: I believe we call this closure only for the `WebView`'s given `url` and not for any further URLs from user-initiated navigations. Should we update the comment to explain that?

itsmeichigo replied:
I don't think so, this closure is triggered by the delegate method `func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!)` which is called when a new page is loaded (redirect success). Can you explain further if I'm misunderstand something here? @selanthiraiyan

itsmeichigo replied:
Good point - updated the comments in 0d62760b81b951e1df147ceee1a483d1da5da8b2

itsmeichigo committed:
Update comments for pageLoadHandler
